### PR TITLE
OOP exercise

### PR DESCRIPTION
**Pull request by rahmanoff**
[Demo](https://rahmanoff.github.io/OOP-exercise/) |
[Code base](https://github.com/rahmanoff/OOP-exercise)

The code is submitted in a dedicated feature branch.

Only code files are submitted.

Please, review.

### CODE REVIEW

**Comment by rahmanoff** [comment]
Hi!
Did morning selfcheck.
Looks ok, as for me.
Pls review.

**Comment by OleksiyRudenko** [comment]
> Hi! Did morning selfcheck. Looks ok, as for me. Pls review.

Double-check yourself. At least 3 requirements aren't met.
Also some approaches from pre-OOP do not work well as they do not scale well. OOP is more than just class syntax.

Feel free asking specific questions to specific requirements items linked by bot if anything is not clear.

**Comment by rahmanoff** [comment]
Hi!
Pls review.
p.s. sorry for bad selfchecks. Do study in parallel with full time job.

**Comment by rahmanoff** [comment]
@OleksiyRudenko , hi!
Any chance to get review?)

File: submissions/Kostiantyn_Rakhmanov/oop-exercise/index.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      return ['species', 'name', 'gender', 'saying'].map(
+         (prop) => `${prop}: ${this[prop]}`
+      ).join('; ');
+   }
+}
+
+class Animal extends Inhabitant {
+   constructor (species, name, gender, saying) {
+      super (species, name, gender, saying);
+      this.legs = 4;
```
Comment (OleksiyRudenko) [inline submissions/Kostiantyn_Rakhmanov/oop-exercise/index.js:20]: Why `name` is generalized to the base class and `legs` are not?
JFYI, methods are just functions, so their parameters can be assigned default values.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@rahmanoff a pretty decent job done.
Just polish it.

**Comment by rahmanoff** [comment]
Hi!
Added 'legs' to Inhabitant.
Pls review.

File: submissions/Kostiantyn_Rakhmanov/oop-exercise/index.js (hunk 1 above)
Comment (OleksiyRudenko) [reply submissions/Kostiantyn_Rakhmanov/oop-exercise/index.js:21]: Why do we need to reassign `legs` here?

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@rahmanoff half-done.
Imagine, you are on an important commercial project.

**Comment by rahmanoff** [comment]
@OleksiyRudenko hi!
Now 'legs' goes to Cat and Dog.

File: submissions/Kostiantyn_Rakhmanov/oop-exercise/index.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+class Dog extends Animal {
+   constructor(name, gender, saying, legs = 4) {
+      super('dog', name, gender, saying, legs);
+   }
+}
+
+class Cat extends Animal {
+   constructor(name, gender, saying, legs = 4) {
+      super('cat', name, gender, saying, legs);
```
Comment (OleksiyRudenko) [inline submissions/Kostiantyn_Rakhmanov/oop-exercise/index.js:32]: Explain what `super` is. It is not a rhetoric question.
This code contains two bugs. They annihilate each other. This is called "transposition of errors".
Can you spot it?
Please note that "it works" is not a good excuse, as bugs tend to reveal themselves during the further development of the project.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@rahmanoff there is a rare case of "transposition of errors" that gives expected outcome at the end but fires back later. Need to figure out and fix.

File: submissions/Kostiantyn_Rakhmanov/oop-exercise/index.js (hunk 2 above)
Comment (rahmanoff) [reply submissions/Kostiantyn_Rakhmanov/oop-exercise/index.js]: Hi! Super consist of all that goes to Parent class.

**Review by OleksiyRudenko — APPROVED** [review]
@rahmanoff great job!

If you rename `printOut` method as `toString` (thus effectively overloading the default `toString` method for an object) then this method will be implicitly called in string contexts (`print` ensures such context).

Details: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/toString

**Comment by rahmanoff** [comment]
@OleksiyRudenko Thank you for your time!)

**Pull request merged**